src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -1379,3 +1529,229 @@ void CustomTextLayout::_OrderRuns()
+    //               +--------------------------------+
+    //               |         advanceWidth           |
+    //
+    //
+    // NOTE: The bearings can be negative, in which case it is specifying that the glyphs overhang the box
+    // as defined by the advanceHeight/width.
+    // See also: https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_glyph_metrics
+    
+    // The scale is a multiplier and the translation is addition. So *1 and +0 will mean nothing happens.
+    float boxVerticalScaleFactor = 1.0f;
+    float boxVerticalTranslation = 0.0f;
+    {
+        // First, find the dimensions of the glyph representing our fully filled box.
+
+        // Ascent is how far up from the baseline we'll draw.
+        // verticalOriginY is the measure from the topLeft corner of the bounding box down to where
+        // the glyph's version of the baseline is.
+        // topSideBearing is how much "gap space" is left between that topLeft and where the glyph
+        // starts drawing. Subtract the gap space to find how far is drawn upward from baseline.
+        const auto boxAscentDesignUnits = boxMetrics.verticalOriginY - boxMetrics.topSideBearing;
+
+        // Descent is how far down from the baseline we'll draw.
+        // advanceHeight is the total height of the drawn bounding box.
+        // verticalOriginY is how much was given to the ascent, so subtract that out.
+        // What remains is then the descent value. Remove the
+        // bottomSideBearing as the "gap space" on the bottom to find how far is drawn downward from baseline.
+        const auto boxDescentDesignUnits = boxMetrics.advanceHeight - boxMetrics.verticalOriginY - boxMetrics.bottomSideBearing;
+
+        // The height, then, of the entire box is just the sum of the ascent above the baseline and the descent below.
+        const auto boxHeightDesignUnits = boxAscentDesignUnits + boxDescentDesignUnits;
+
+        // Second, find the dimensions of the cell we're going to attempt to fit within.
+        // We know about the exact ascent/descent units in pixels as calculated when we chose a font and
+        // adjusted the ascent/descent for a nice perfect baseline and integer total height.
+        // All we need to do is adapt it into Design Units so it meshes nicely with the Design Units above.
+        // Use the formula: Pixels * Design Units Per Em / Font Size = Design Units
+        const auto cellAscentDesignUnits = ascentPixels * fontMetrics.designUnitsPerEm / scaledFontSize;
+        const auto cellDescentDesignUnits = descentPixels * fontMetrics.designUnitsPerEm / scaledFontSize;
+        const auto cellHeightDesignUnits = cellAscentDesignUnits + cellDescentDesignUnits;
+
+        // OK, now do a few checks. If the drawn box touches the top and bottom of the cell
+        // and the box is overall tall enough, then we'll not bother adjusting.
+        // We will presume the font author has set things as they wish them to be.

```

Comment:
aside: which fonts does this work for?

Reply:
(mostly for my own edification: I want to know which fonts just pass through this without a single adjustment)

Reply:
Oh, the same ones as I described in the original diagram. The ones where I had "3 or 4 green boxes" in my spreadsheet.

- Cascadia Patched
- Envy Code R
- Fira Code
(and without the gap fix, monofur italic)